### PR TITLE
Solution to trait alias bug #65673

### PR DESCRIPTION
**Pull request by alexreg**
Supercedes PR #66392.

I believe the correct thing to do is to expand the trait aliases to check everything is in order (single non-auto trait, etc.) but then use the unexpanded list of trait to construct the existential predicates for the trait object.

r? @eddyb

CC @estebank @nikomatsakis @Centril

### CODE REVIEW

**Comment by alexreg** [comment]
@estebank Any idea what's going on with the diagnostics changes above?

File: src/librustc_typeck/astconv.rs [hunk 1]
```diff
@@ -12,32 +11,34 @@ use crate::lint;
...
 use std::collections::BTreeSet;
 use std::iter;
 use std::slice;
 
-use rustc_data_structures::fx::FxHashSet;
-
-use rustc_error_codes::*;
-
 #[derive(Debug)]
 pub struct PathSeg(pub DefId, pub usize);
```
Comment (eddyb) [inline src/librustc_typeck/astconv.rs]: Please do not include unrelated changes, it pointlessly makes review hard (if you use `git gui` you can selectively commit some changes and discard the rest, for example).

File: src/librustc_typeck/collect.rs [hunk 2]
```diff
@@ -39,16 +45,9 @@ use syntax::feature_gate;
...
-use rustc::hir::Node;
-use rustc::hir::def_id::{DefId, LOCAL_CRATE};
-use rustc::hir::intravisit::{self, NestedVisitorMap, Visitor};
-use rustc::hir::GenericParamKind;
-use rustc::hir::{self, CodegenFnAttrFlags, CodegenFnAttrs, Unsafety};
-
 use errors::{Applicability, StashKey};
-
 use rustc_error_codes::*;
+use rustc_target::spec::abi;
```
Comment (eddyb) [inline src/librustc_typeck/collect.rs]: Same here.

File: src/librustc_typeck/collect.rs [hunk 3]
```diff
@@ -729,16 +728,15 @@ fn super_predicates_of(
...
     // In the case of trait aliases, however, we include all bounds in the where-clause,
-    // so e.g., `trait Foo = where u32: PartialEq<Self>` would include `u32: PartialEq<Self>`
+    // so, e.g., `trait Foo = where u32: PartialEq<Self>` would include `u32: PartialEq<Self>`
     // as one of its "superpredicates".
     let is_trait_alias = tcx.is_trait_alias(trait_def_id);
     let superbounds2 = icx.type_parameter_bounds_in_generics(
         generics, item.hir_id, self_param_ty, OnlySelfBounds(!is_trait_alias));
 
-    // Combine the two lists to form the complete set of superbounds:
+    // Combine the two lists to form the complete set of superbounds.
```
Comment (eddyb) [inline src/librustc_typeck/collect.rs]: None of this file is effectively changed, please do not include it in the PR.

File: src/librustc/traits/object_safety.rs [hunk 4]
```diff
@@ -462,7 +462,7 @@ impl<'tcx> TyCtxt<'tcx> {
             .filter(|(_, item)| item.kind == ty::AssocKind::Type)
             .collect::<Vec<_>>();
 
-        // existential predicates need to be in a specific order
+        // Existential predicates need to be in a specific order.
```
Comment (eddyb) [inline src/librustc/traits/object_safety.rs]: Same here, please do not commit formatting changes, especially in files that are otherwise unchanged.

File: src/librustc/session/mod.rs [hunk 5]
```diff
@@ -332,7 +331,7 @@ impl Session {
             None => self.warn(msg),
         }
     }
-    /// Delay a span_bug() call until abort_if_errors()
+    /// Delay a `span_bug()` call until `abort_if_errors()`.
```
Comment (eddyb) [inline src/librustc/session/mod.rs]: Same here.

File: src/librustc/traits/object_safety.rs (hunk 4 above)
Comment (alexreg) [reply src/librustc/traits/object_safety.rs]: “Pointless” is your opinion, but you’re right, they are not too close to functional changes so I’ll remove them.

**Comment by alexreg** [comment]
@eddyb I was hoping for a review of the actual functionality changes here or a suggestion as to why the diagnostics changed, but if you feel you’re not qualified to review, I can r? Niko or someone else. Just let me know, ta.

**Comment by eddyb** [comment]
I cannot easily see where the functional changes are when 90% of the non-test changes aren't that. Anyone else who doesn't point that out still wasted time searching for the functional changes.

**Comment by oli-obk** [comment]
Please remove all nonfunctional changes. It requires a lot of time for reviewers to find the functional changes in your PR.

Please do not write suggestive comments about a reviewer's ability to review your PRs. Not only did you choose them yourself, so you thought them qualified, but the desire not to review was rooted in the fact that it just requires a lot of time that both you and the reviewer could have better spent.

r? @oli-obk

File: src/librustc/session/mod.rs (hunk 5 above)
Comment (alexreg) [reply src/librustc/session/mod.rs]: Fair

File: src/librustc/traits/object_safety.rs (hunk 4 above)
Comment (alexreg) [reply src/librustc/traits/object_safety.rs]: Done

File: src/librustc_typeck/astconv.rs (hunk 1 above)
Comment (alexreg) [reply src/librustc_typeck/astconv.rs]: Well at least I'm changing this file... seems like too minor a quibble here.

**Comment by BurntSushi** [comment]
Moderator note: please do not abuse the moderation features on github. If anyone isn't clear on what constitutes abuse, then please reach out to rust-mods@rust-lang.org.

**Comment by alexreg** [comment]
It was not certainly not abuse of the feature. Rather, I marked the comment as abuse (I consider it both patent harassment given his history), and did not want it to derail more germane discussion. Furthermore, it the first paragraph repeated a point that had already been made quite forcefully, and the second was a presumptive and rude comment that should not have been made in public.

**Comment by BurntSushi** [comment]
Moderator note: this is not the place to discuss moderation. As I said in my previous comment, please email the mods if you would like to discuss this further. Off topic comments in this thread will continue to be deleted.

**Comment by alexreg** [comment]
Looking into the diagnostics issue now, but pending that, I think this is the right solution.

(In case no one has realised yet, non-functional changes that were accidentally left in following reverts to several files were removed shortly after the above discussion.)

**Comment by alexreg** [comment]
So, the effects on diagnostics for a few compile-fail tests aren't actually that bad, it seems (having discussed them briefly with @Centril). In fact, in one way (the inclusion of an extra diagnostic when you try to use two non-auto-traits), it seems to be better. Three ui tests are affected. So, @estebank, if you feel otherwise, do please let me know and we can try to improve them either here or in another PR. Otherwise, I'll leave this PR to Niko to review in his own time.

File: src/librustc_typeck/astconv.rs [hunk 6]
```diff
@@ -1247,8 +1247,10 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
         // Expand trait aliases recursively and check that only one regular (non-auto) trait
         // is used and no 'maybe' bounds are used.
         let expanded_traits =
-            traits::expand_trait_aliases(tcx, bounds.trait_bounds.iter().cloned());
-        let (mut auto_traits, regular_traits): (Vec<_>, Vec<_>) =
+            traits::expand_trait_aliases(tcx, bounds.trait_bounds.iter().cloned())
+                // Ensure that trait ref is to self type and not some type param.
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/librustc_typeck/astconv.rs]: I'm a bit worried here about an example like this:

```rust
#![feature(trait_alias)]

trait Foo<T> = where T: PartialEq<Self>;

fn main() {
    let _: &dyn Foo<()> = &();
}
```

When I test this, I currently get some wacky ICE =) But it seems like we would screen out the `T: PartialEq<Self>` from the list of `regular_traits` here. Then, later, when we check for object safety, we will never see that `T: PartialEq<Self>` is not object-safe. 

My intuition is that `trait Foo<T> = ...` should work exactly as if you created a trait and implemented it, for the most part. And [if we do that](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=8dab4b2fc3984d04ee5f38f5b0c87f5d) with this example we get object safety errors. I don't quite see where they are supposed to crop up here.

File: src/librustc_typeck/astconv.rs [hunk 7]
```diff
@@ -1398,50 +1401,53 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
         // Calling `skip_binder` is okay because the predicates are re-bound.
-        let regular_trait_predicates = existential_trait_refs.map(
-            |trait_ref| ty::ExistentialPredicate::Trait(*trait_ref.skip_binder()));
-        let auto_trait_predicates = auto_traits.into_iter().map(
-            |trait_ref| ty::ExistentialPredicate::AutoTrait(trait_ref.trait_ref().def_id()));
+        let regular_trait_predicates = existential_trait_refs
+            .map(|trait_ref| ty::ExistentialPredicate::Trait(*trait_ref.skip_binder()));
+        let auto_trait_predicates = auto_traits
+            .into_iter()
+            .map(|trait_ref| ty::ExistentialPredicate::AutoTrait(trait_ref.def_id()));
```
Comment (nikomatsakis) [inline src/librustc_typeck/astconv.rs]: To confirm, also no functional changes here, correct? ( I didn't spot any )

**Review by nikomatsakis — COMMENTED** [review]
I think I understand what this PR is going for, and it mostly makes sense, but I don't quite think I grasp how we should be doing object safety checks on trait aliases. I left some comments, maybe @alexreg you have thoughts.

I'm largely sidestepping the question about cosmetic changes. I'd rather discuss it offline. However, I do think that the general team stance (per our prior discussions) that is that unnecessary cosmetic changes should be minimized. I did leave a few comments to verify that some sets of changes were indeed non-functional as they appeared to be; I can attest it took me some time comparing the code to check that this was the case.

File: src/librustc_typeck/astconv.rs (hunk 7 above)
Comment (alexreg) [reply src/librustc_typeck/astconv.rs]: Correct, just made it more readable since I was editing this fn anyway.

File: src/librustc/traits/util.rs [hunk 8]
```diff
@@ -358,19 +358,21 @@ impl<'tcx> TraitAliasExpander<'tcx> {
         debug!("expand_trait_aliases: trait_ref={:?}", trait_ref);
 
         // Don't recurse if this bound is not a trait alias.
-        let is_alias = tcx.is_trait_alias(trait_ref.def_id());
-        if !is_alias {
+        if !tcx.is_trait_alias(trait_ref.def_id()) {
```
[the remark this answers is not in the document]
Comment (alexreg) [reply src/librustc/traits/util.rs]: Yeah, I can remove this change if you like. I think I just forgot. Makes sense to inline this, but it's not important.

File: src/librustc_typeck/astconv.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexreg) [reply src/librustc_typeck/astconv.rs]: The key thing is to avoid trying to create existential predicates for predicates where the self ty is not the dummy self ty, which is of course impossible. This filter might be blocking out *too* much (projections on `Self` too?), but all the tests seem to be working anyway...

The relevant test is `issue-65673.rs`, I believe (pre-existing, slightly modified).

[the remark this answers is not in the document]
Comment (alexreg) [reply src/librustc_typeck/astconv.rs]: `trait Foo = where Self: Display;` should (and I think is) equivalent to `trait Foo = Display;` in most circumstances, except as a trait object. Since technically no principal trait is specified, the compiler will complain. Perhaps we should do that, but perhaps we should instead check that there's at least one predicate instead of trait?

[the remark this answers is not in the document]
Comment (alexreg) [reply src/librustc_typeck/astconv.rs]: As for object safety, I guess we need to modify that code explicitly? Given such predicates on `T` definitely cannot be included in the trait object itself (they're simply not existential).

File: src/librustc_typeck/astconv.rs [hunk 9]
```diff
@@ -1398,50 +1401,53 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
             err.emit();
         }
 
+        let (mut auto_traits, regular_traits): (Vec<_>, Vec<_>) =
```
[the remark this answers is not in the document]
Comment (alexreg) [reply src/librustc_typeck/astconv.rs]: Yep, this fn was already very big, and that’s just gotten worse now. I’ll see what I can do.

**Comment by hdhoang** [comment]
ping from triage, @nikomatsakis any further review after the author's comment?

**Comment by JohnCSimon** [comment]
Ping from triage: @alexreg can you please address the merge conflicts?

**Comment by alexreg** [comment]
@JohnCSimon Happy to rebase and fix conflicts, but it would be nice to get the node from Niko that this approach is the one we want to go with (per his last comments) before doing so. Ta.

File: src/librustc_typeck/astconv.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/librustc_typeck/astconv.rs]: I don't quite understand your response here, @alexreg. 

What is the behavior on the example I gave, for example? I don't think that `issue-65673.rs` is equivalent. For one thing, it's run-pass, but I expect an error, but also the trait alias in `issue-65673.rs` is:

```rust
 trait Alias<T> = Any where T: Trait; 
```

whereas the trait alias in my example was `trait Foo<T> = where T: PartialEq<Self>;`, which (notably) references `Self` (unlike `issue-65673.rs`).

Also, regarding the difference between `trait Foo = Display` and `trait Foo where Self: Display`, I guess I'd have to consult the RFC. It's a *touch* surprising to me since `trait Foo: Display` and `trait Foo where Self: Display` *are* equivalent, for better or worse.

**Comment by joelpalmer** [comment]
Triaged

File: src/librustc_typeck/astconv.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (alexreg) [reply src/librustc_typeck/astconv.rs]: Ah, `complain_about_missing_associated_types` has already been factored out now, which makes this fn size much more reasonable!

File: src/librustc_typeck/astconv.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexreg) [reply src/librustc_typeck/astconv.rs]: @nikomatsakis Okay, I've gotten your expected behaviour implemented. (And I managed to remove the `Any` from `issue-65673.rs`.) Now, the question is, do we want to allow just `dyn` or `dyn 'static` as a type?

[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/librustc_typeck/astconv.rs]: I don't know! I think I'm inclined to preserve the status quo, whatever it is, though if/when we generalize to `dyn A + B` I'd also be inclined to accept zero bounds.

[the remark this answers is not in the document]
Comment (alexreg) [reply src/librustc_typeck/astconv.rs]: @nikomatsakis Okay, sounds fair. I'll push shortly and this PR will be ready for review.

**Comment by joelpalmer** [comment]
Closing due to inactivity. Please reopen when changes are made.
